torch/package/package_importer.py

```diff
@@ -381,7 +381,11 @@ def _do_find_and_load(self, name):
                 return self.modules[name]
             parent_module = self.modules[parent]
             try:
-                path = parent_module.__path__  # type: ignore[attr-defined]
+                # Extern check is to handle edge case described in:
```

Comment:
I think a more restricted special case would be better. In general, being an extern module should not exempt you from this check. The problem arises because `os` assumes that populating `sys.modules` is sufficient to get importing to work.

What about: when importing `os` in `_gcd_import`, add a special case that adds `os.path` to `self.modules`, essentially replicating the hackery `os` does.

Reply:
Do we want to just solve this for `os`/`os.path`? The problem is more generalized: https://github.com/pytorch/pytorch/issues/57490#issuecomment-864159922

Reply:
Yeah, the problem is more general, but disabling the package check for all extern modules is too broad a fix. In most cases we should still expect things with submodules to be packages, regardless of whether they are extern or not.

The real source of the problem is just the mismatch between sys.modules and self.modules. A targeted fix would detect that the import of a module has changed sys.modules. But that's probably too tricky relative to the number of modules that actually do this, so my suggestion is to just hard code a fix for `os`. 

I think we should not expect the general version of hacking `sys.modules` to work, but we should make a concession to supporting the places the standard library does it in the name of supporting standard library usage transparently.

Reply:
Okay- I can change this to a hard code fix. Thanks for looking at it!